packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,245 @@
+  countApprovedOptional: () => 0,
+  areGroupConsentsApproved: () => false,
+});
+
+export const getConsentsFromConsentGroup = (groups: ConsentGroup[]): ConsentList => {
+  return groups.reduce((ids, currentGroup) => {
+    currentGroup.consents.forEach((consentData) => {
+      ids.push(consentData.id);
+    });
+    return ids;
+  }, []);
+};
+
+export const Provider = ({ cookieDomain, children, content }: CookieConsentContextProps): React.ReactElement => {
+  const requiredConsents = content.requiredConsents
+    ? getConsentsFromConsentGroup(content.requiredConsents.groupList)
+    : undefined;
+  const optionalConsents = content.optionalConsents
+    ? getConsentsFromConsentGroup(content.optionalConsents.groupList)
+    : undefined;
+  const consentController = useMemo(() => create({ requiredConsents, optionalConsents, cookieDomain }), []);
+  const hasUserHandledAllConsents = () =>
+    consentController.getRequiredWithoutConsent().length === 0 && consentController.getUnhandledConsents().length === 0;
+
+  const mergeConsents = () => ({
+    ...consentController.getRequired(),
+    ...consentController.getOptional(),
+  });
+
+  const [, forceUpdate] = useState<number>(0);
+  const reRender = () => {
+    forceUpdate((p) => p + 1);
+  };
+
+  const save = () => {
+    const savedData = consentController.save();
+    if (hasUserHandledAllConsents()) {
+      reRender();
+      if (content.onAllConsentsGiven) {
+        content.onAllConsentsGiven(mergeConsents());
+      }
+    }
+    return savedData;
+  };
+
+  const update: CookieConsentContextType['update'] = (key, value) => {
+    consentController.update(key, value);
+  };
+  const approveSelectedAndRequired = () => {
+    consentController.approveRequired();
+    save();
+  };
+  const approveAll: CookieConsentContextType['approveAll'] = () => {
+    consentController.approveAll();
+    save();
+  };
+  const approveRequired: CookieConsentContextType['approveRequired'] = () => {
+    Object.keys(consentController.getOptional()).forEach((optionalConsent) => {
+      update(optionalConsent, false);
+    });
+    consentController.approveRequired();
+    save();
+  };
+  const setOptional = (approved: boolean) => {
+    Object.keys(consentController.getOptional()).forEach((optionalConsent) => {
+      update(optionalConsent, approved);
+    });
+  };
+
+  const onAction: CookieConsentContextType['onAction'] = (action, consents, value) => {
+    if (action === 'approveAll') {
+      approveAll();
+    } else if (action === 'approveRequired') {
+      approveRequired();
+    } else if (action === 'approveSelectedAndRequired') {
+      approveSelectedAndRequired();
+    } else if (action === 'changeConsentGroup') {
+      consents.forEach((consent) => {
+        update(consent, value);
+      });
+    } else if (action === 'approveOptional') {
+      setOptional(true);
+    } else if (action === 'unapproveOptional') {
+      setOptional(false);
+    }
+    reRender();
```

Comment:
Is this `reRender` necessary? There seems to be another `reRender` that is called at the end of almost each of the if clauses. For example, `approveAll` will call function `save`, and if user has handled all consents, the `save` function will call `reRender`. So reRender gets called twice, I wonder if it's always necessary?